### PR TITLE
Room Scale HMD comfort features.

### PR DESCRIPTION
**Pull request by hyperlogic**
The goal of this PR is to increase comfort for HMD users, by eliminating instances where we move the HMD camera unexpectedly.

When walking in room scale you are no longer pushed by collision.  Your avatar will attempt to follow your HMD camera position in a physical way.  If your body cannot reach your HMD camera, because it is blocked by collision, you will separate from your body and the hand controller balls will no longer match your avatar's hands.  However, you can still grab and manipulate objects with the controller balls.  You can rejoin your body by walking back into it, teleporting, or moving out of collision in such a way that it can walk back to you.

Other features include:
- "Advanced Movement For Hand Controllers" in Settings menu, now drives the HMD with linear velocity and without collision.
- Snappier teleport behavior for HMD users.
- The avatar's hips will no longer embed into collision.
- The avatar collision can now drive over small bumps and obstructions.
- Avatar drive speed has been reduced to 2 m/s.
- Avatar no longer plays walk animation when motion is impeded by collision.
- Avatar walk animations for large or small avatars should have less foot sliding.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: eccd01e88f3f5b0b39307a7e967d0c4fb5895809 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-eccd01e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-eccd01e.dmg

**Comment by cozza13** [comment]
**QA:**

- General smoke test for performance
- Go to dev-demo and test distance grabbing and equipping. Make sure it is not a degraded performance. You should always be able to move you HMD camera view without running into virtual obstacles

**PR Specific (Need HMD and hand controllers. Ideally Vive or Oculus touch)**

- Go to dev-demo with HMD and hand controllers.
- Go to the dice table and grab and throw the dice. You should be able to lean into the table without your camera being pushed back (in the current release your camera would get pushed back)
- While you are at the Dice table and in room scale, physically walk yourself through the table. Turn around and look back. You should see that your camera has detached from your head and your avatar is standing at the table. 
  - Confirm that the avatar position looks good (It is not trying to walk, it is not buckled... ) take a screen shot of your avatar and attach it in you comments. 
- While still inside the table, physically move yourself outside of the table to a place that your avatar could get to you. Your avatar should walk back into your body.
- Teleport around the room and note that your avatar will get back into you body. 
- Go up to a wall and physically walk through it, confirm that you are able to look through the wall without you HMD view being pushed back
  - with another person watching, confirm that your avatar is not trying to climb up the wall. 
- In settings, enable 'Advanced Movement For Hand Controllers' this will allow you to drive. You should be able to drive your camera without any collisions stopping it. 
- Use some of the items from the marketplace. e.g. https://highfidelity.io/marketplace/items/de2f7e5c-4945-48a0-a045-fe602d7fa4b9 and confirm they are working

**Comment by hifi-gustavo** [comment]
Build console output for commit: aab974e6ae1e58d4dcad01abeefcec7ea8434dc3 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-aab974e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-aab974e.dmg

**Comment by thoys** [comment]
QA:

General smoke test for performance
Go to dev-demo and test distance grabbing and equipping. Make sure it is not a degraded performance. You should always be able to move you HMD camera view without running into virtual obstacles
PR Specific (Need HMD and hand controllers. Ideally Vive or Oculus touch)

Go to dev-demo with HMD and hand controllers.
Go to the dice table and grab and throw the dice. You should be able to lean into the table without your camera being pushed back (in the current release your camera would get pushed back)
While you are at the Dice table and in room scale, physically walk yourself through the table. Turn around and look back. You should see that your camera has detached from your head and your avatar is standing at the table.
Confirm that the avatar position looks good (It is not trying to walk, it is not buckled... ) take a screen shot of your avatar and attach it in you comments.
While still inside the table, physically move yourself outside of the table to a place that your avatar could get to you. Your avatar should walk back into your body.
Teleport around the room and note that your avatar will get back into you body.
Go up to a wall and physically walk through it, confirm that you are able to look through the wall without you HMD view being pushed back
with another person watching, confirm that your avatar is not trying to climb up the wall.
In settings, enable 'Advanced Movement For Hand Controllers' this will allow you to drive. You should be able to drive your camera without any collisions stopping it.
Use some of the items from the marketplace. e.g. https://highfidelity.io/marketplace/items/de2f7e5c-4945-48a0-a045-fe602d7fa4b9 and confirm they are working

**Comment by hyperlogic** [comment]
There's one bug we noticed with this PR, if you drive over a bump, it tends to throw your avatar up in the air, like you jump off the edge of the bump.   This is something @AndrewMeadows is looking into and should have a fix for soon.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 895b4e3dcc3024eecc400c31c68b9758a2781fe1 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-895b4e3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-895b4e3.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: cc7a6ada38a4b14e419d949c5414566f09f21acb available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-cc7a6ad.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-cc7a6ad.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b0c7564b604bba17a0bf2191fc471c8e9b108c10 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-b0c7564.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-b0c7564.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2fdd5e43680ca1184ee005a19a03083d89501528 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-2fdd5e4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-2fdd5e4.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 55c90e7bca86b57f678d9eea6c3905747cbbbf65 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-55c90e7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-55c90e7.dmg

**Comment by AndrewMeadows** [comment]
BTW, I tested teleport.js after the last merge (the conflicts were in that file) and verified that it still works and is functional.

**Comment by hyperlogic** [comment]
Additional QA test plan item:

1. Login with Vive while standing
2. Navigate to Welcome domain
3. Sit on floor
4. Press ESC twice. The first press goes into PAUSED mode, the second press unpauses and recenters the avatar so that its eyes line up with the known HMD position (which happens to be inside or below the floor).
--> on bugged versions the avatar's physical body will end up in deep penetration with the floor and will be resolved out, often out the bottom.
--> on fixed version the avatar's physical body will NOT be dropped... it the avatar was standing on the ground when the recenter happens. If the avatar is hovering then the recenter will properly move the avatar's physical body down accordingly.

**Comment by hifi-gustavo** [comment]
Build console output for commit: d54c07ed14a2f1156cc3db8cc8e1a71e25155df2 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-d54c07e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-d54c07e.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 81db8f7e086e6f011638100badfc950da40e67a7 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-81db8f7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-81db8f7.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 35aebd21e661559ca4fd5ea49be5c73004bacfc4 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-35aebd2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-35aebd2.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c8ad52c5875d46279f6b41627313594e5b8bee6f available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-c8ad52c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-c8ad52c.dmg

**Comment by thoys** [comment]
can you update this to the latest master please @hyperlogic ?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 51a9a99381e6bed02c7fe770a203f3826fac6d57 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-51a9a99.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-51a9a99.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c78f2b5409aa94d9b573e19d4bc8bf78ec1fec0c available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-c78f2b5.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-c78f2b5.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d86242ae74fb6ae3b75ab4048300253290c6df93 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-d86242a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-d86242a.dmg

**Comment by RyanDowne** [comment]
I just tested this in dev-welcome and it worked great. I was unable to walk through walls but I was able to walk up wedges that were dynamic or static ramps. I did not see any drop in framerate in dev-welcome.

**Comment by RyanDowne** [comment]
Here is interesting result of this PR. If you go in the HMD to one location, switch to desktop and fly your avatar to another then click the HMD, you end up back where you started. It's very confusing but I am guessing by design somehow.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 38e4ac0ae9b4ce70e7b94cdf581d3e5d66f2be15 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-38e4ac0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-38e4ac0.dmg

**Comment by AndrewMeadows** [comment]
The jump back to old position when changing back to HMD- out of Desktop-mode is a bug.

**Comment by RyanDowne** [comment]
This PR also has a bug where you can never fly. The upside to this was that we were able to fling ourselves into the atmosphere by standing on the edge of a teetering platform and dropping a heavy block on the other end. Jimi has a video if you wanna see. We need to figure out a way to make both true -- when you click up on the right stick you can fly but if you get flung it doesn't auto put you in fly mode.

**Comment by hifi-gustavo** [comment]
Build console output for commit: ccdf3831513eb0313c8d80f81b78f0c28393a766 available at: https://gist.github.com/da820841a16520227c668d7b2e951a8c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-ccdf383.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-ccdf383.dmg

**Comment by cozza13** [comment]
ok to test

**Comment by hifi-gustavo** [comment]
Build console output for commit: ccdf3831513eb0313c8d80f81b78f0c28393a766 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-ccdf383.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-ccdf383.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5f0761598e5786e8d75c856bb55c46aa40e2c307 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-5f07615.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-5f07615.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6bb3dd6830824f9e133572cd9e958c6b0c08de77 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-6bb3dd6.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-6bb3dd6.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1625430e0e271f87454737dcaca89fe72213de45 available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-1625430.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-1625430.dmg

File: interface/src/avatar/MyCharacterController.cpp [hunk 1]
```diff
@@ -15,11 +15,17 @@
...
 #include "MyAvatar.h"
 
-// TODO: improve walking up steps
-// TODO: make avatars able to walk up and down steps/slopes
 // TODO: make avatars stand on steep slope
 // TODO: make avatars not snag on low ceilings
 
+
+void MyCharacterController::RayShotgunResult::reset() {
+    //hitNormal = glm::vec3(0.0f);
```
Comment (sethalves) [inline interface/src/avatar/MyCharacterController.cpp]: cruft?

File: interface/src/avatar/MyCharacterController.cpp [hunk 2]
```diff
@@ -15,11 +15,17 @@
...
+    hitFraction = 1.0f;
+    //numHits = 0;
```
Comment (sethalves) [inline interface/src/avatar/MyCharacterController.cpp]: cruft?

File: interface/src/avatar/MyCharacterController.cpp [hunk 3]
```diff
@@ -72,12 +74,419 @@ void MyCharacterController::updateShapeIfNecessary() {
...
+        btScalar stepX = 2.0f * (_radius - slop) / (btScalar)(NUM_COLUMNS - 1);
+
+        btTransform transform = _rigidBody->getWorldTransform();
+        btVector3 position = transform.getOrigin();
+        btMatrix3x3 rotation = transform.getBasis();
+
+        for (int32_t i = 0; i < NUM_ROWS; ++i) {
+            int32_t maxJ = NUM_COLUMNS;
+            btScalar offsetX = -(btScalar)((NUM_COLUMNS - 1) / 2) * stepX;
+            if (i%2 == 1) {
```
Comment (sethalves) [inline interface/src/avatar/MyCharacterController.cpp]: space around operator

File: interface/src/avatar/MyCharacterController.cpp [hunk 4]
```diff
@@ -72,12 +74,419 @@ void MyCharacterController::updateShapeIfNecessary() {
...
```
Comment (sethalves) [inline interface/src/avatar/MyCharacterController.cpp]: space around operator

File: interface/src/avatar/MyCharacterController.h [hunk 5]
```diff
@@ -24,10 +24,39 @@ class MyCharacterController : public CharacterController {
...
+    // Sweeping a convex shape through the physics simulation can expensive when the obstacles are too complex
+    // (e.g. small 20k triangle static mesh) so instead as a fallback we cast several rays forward and if they
+    // don't hit anything we consider it a clean sweep.  Hence the "Shotgun" code.
+    class RayShotgunResult {
+    public:
+        void reset();
+
+        //glm::vec3 hitNormal { 0.0f, 0.0f, 0.0f };
+        float hitFraction { 1.0f };
+        //int32_t numHits { 0 };
```
Comment (sethalves) [inline interface/src/avatar/MyCharacterController.h]: cruft?

File: libraries/physics/src/CharacterController.h [hunk 6]
```diff
@@ -155,21 +168,29 @@ class CharacterController : public btCharacterControllerInterface {
     quint32 _jumpButtonDownCount;
     quint32 _takeoffJumpButtonID;
 
-    btScalar _halfHeight;
-    btScalar _radius;
+    // data for walking up steps
+    btVector3 _stepPoint;
```
Comment (sethalves) [inline libraries/physics/src/CharacterController.h]: no initializer for this one?

File: libraries/physics/src/CharacterGhostObject.cpp [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+
+        if (!collisionPair->m_algorithm) {
+            // null m_algorithm means the two shape types don't know how to collide!
+            // shouldn't fall in here but just in case
+            continue;
+        }
+
+        btScalar mostFloorPenetration = 0.0f;
+        collisionPair->m_algorithm->getAllContactManifolds(manifoldArray);
+        for (int j = 0;j < manifoldArray.size(); j++) {
```
Comment (sethalves) [inline libraries/physics/src/CharacterGhostObject.cpp]: space after ;

File: libraries/physics/src/CharacterGhostObject.h [hunk 8]
```diff
@@ -0,0 +1,103 @@
+//
+//  CharacterGhostObject.h
+//  libraries/physcis/src
```
Comment (sethalves) [inline libraries/physics/src/CharacterGhostObject.h]: physcis

File: libraries/physics/src/CharacterGhostShape.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
+//
+//  CharacterGhostShape.h
+//  libraries/physcis/src
```
Comment (sethalves) [inline libraries/physics/src/CharacterGhostShape.h]: physcis

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2bbe2be51649049a9e6b150b1a4e8ec1bd9f41da available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-2bbe2be.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-2bbe2be.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ae699e2a654e60bbfd6f60a3ba88328e32f18e5f available at: https://gist.github.com/927d87f99c0d54aa0902760ad582cefc<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-ae699e2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8691-ae699e2.dmg

**Comment by OmegaHeron** [comment]
An issue I'm running into while using release build based on this (5749) is as follows;
This is in Desktop mode, Windows 10

I begin interface session at some arbitrary position in domain
I move some arbitrary distance then I reset my sensors using `'`
I am moved back to where my session began (or if I've changed domains - to point I began new domain session).

Starting at this view - note my position;

hifi-snap-by-omegaheron-on-2016-12-01_22-37-55

Moving away a bit - again note position;

hifi-snap-by-omegaheron-on-2016-12-01_22-38-45

Hit `'` to reset sensors and I'm moved back to;

hifi-snap-by-omegaheron-on-2016-12-01_22-39-18

**Pull request merged**